File: src/main/java/earth/terrarium/tempad/common/utils/CodecUtils.java
package earth.terrarium.tempad.common.utils;

import com.teamresourceful.bytecodecs.base.ByteCodec;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Registry;
import net.minecraft.core.registries.Registries;
import net.minecraft.resources.ResourceKey;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.level.Level;

public class CodecUtils {
    public static final ByteCodec<ResourceLocation> RESOURCE_LOCATION = ByteCodec.STRING.map(ResourceLocation::new, ResourceLocation::toString);
    public static final ByteCodec<ResourceKey<Level>> DIMENSION = resourceKeyByteCodec(Registries.DIMENSION);
    public static final ByteCodec<BlockPos> BLOCK_POS = ByteCodec.LONG.map(BlockPos::of, BlockPos::asLong);

    public static <T> ByteCodec<ResourceKey<T>> resourceKeyByteCodec(ResourceKey<? extends Registry<T>> registry) {
        return RESOURCE_LOCATION.map(resourceLocation -> ResourceKey.create(registry, resourceLocation), ResourceKey::location);
    }
}

File: src/main/kotlin/earth/terrarium/tempad/common/utils/CodecUtils.kt
package earth.terrarium.tempad.common.utils

import com.mojang.authlib.GameProfile
import com.mojang.serialization.Codec
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import com.teamresourceful.bytecodecs.base.ByteCodec
import com.teamresourceful.bytecodecs.base.`object`.ObjectByteCodec
import com.teamresourceful.resourcefullib.common.bytecodecs.StreamCodecByteCodec
import com.teamresourceful.resourcefullib.common.color.Color
import net.minecraft.core.UUIDUtil
import net.minecraft.world.phys.Vec3
import net.neoforged.neoforge.fluids.crafting.FluidIngredient
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

val COLOR_BYTE_CODEC: ByteCodec<Color> = ByteCodec.BYTE.dispatch(
    { aByte: Byte ->
        when (aByte) {
            0.toByte() -> ByteCodec.unit(Color.DEFAULT)
            1.toByte() -> ByteCodec.unit(Color.RAINBOW)
            else -> ByteCodec.INT.map(::Color, Color::getValue)
        }
    },
    { color: Color ->
        when {
            color.isDefault -> 0.toByte()
            color.isRainbow -> 1.toByte()
            else -> 2.toByte()
        }
    })

val VEC3_BYTE_CODEC: ByteCodec<Vec3> = ObjectByteCodec.create(
    ByteCodec.DOUBLE.fieldOf { it.x },
    ByteCodec.DOUBLE.fieldOf { it.y },
    ByteCodec.DOUBLE.fieldOf { it.z },
    ::Vec3
)

val FLUID_INGREDIENT_BYTE_CODEC: ByteCodec<FluidIngredient> = StreamCodecByteCodec.ofRegistry(FluidIngredient.STREAM_CODEC)

val DATE_CODEC: Codec<Date> = Codec.STRING.xmap(String::toLong, Long::toString).xmap(::Date, Date::getTime)

val DATE_BYTE_CODEC: ByteCodec<Date> = ByteCodec.LONG.map(::Date, Date::getTime)

val GAME_PROFILE_BYTE_CODEC: ByteCodec<GameProfile> = ObjectByteCodec.create(
    ByteCodec.UUID.fieldOf { it.id },
    ByteCodec.STRING.fieldOf { it.name },
    ::GameProfile
)

val GAME_PROFILE_CODEC: MapCodec<GameProfile> = RecordCodecBuilder.mapCodec {
    it.group(
        UUIDUtil.CODEC.fieldOf("id").forGetter { it.id },
        Codec.STRING.fieldOf("name").forGetter { it.name },
    ).apply(it, ::GameProfile)
}